crates/rover-client/src/query/supergraph/fetch.rs

```diff
@@ -0,0 +1,139 @@
+use crate::blocking::StudioClient;
+use crate::RoverClientError;
+use graphql_client::*;
+
+// I'm not sure where this should live long-term
+/// this is because of the custom GraphQLDocument scalar in the schema
+type GraphQLDocument = String;
+
+#[derive(GraphQLQuery)]
+// The paths are relative to the directory where your `Cargo.toml` is located.
+// Both json and the GraphQL schema language are supported as sources for the schema
+#[graphql(
+    query_path = "src/query/supergraph/fetch.graphql",
+    schema_path = ".schema/schema.graphql",
+    response_derives = "PartialEq, Debug, Serialize, Deserialize",
+    deprecated = "warn"
+)]
+/// This struct is used to generate the module containing `Variables` and
+/// `ResponseData` structs.
+/// Snake case of this name is the mod name. i.e. fetch_supergraph_query
+pub struct FetchSupergraphQuery;
+
+/// The main function to be used from this module. This function fetches a
+/// core schema from apollo studio
+pub fn run(
+    variables: fetch_supergraph_query::Variables,
+    client: &StudioClient,
+) -> Result<String, RoverClientError> {
+    let graph = variables.graph_id.clone();
+    let invalid_variant = variables.variant.clone();
+    let response_data = client.post::<FetchSupergraphQuery>(variables)?;
+    get_supergraph_sdl_from_response_data(response_data, graph, invalid_variant)
+}
+
+fn get_supergraph_sdl_from_response_data(
+    response_data: fetch_supergraph_query::ResponseData,
+    graph: String,
+    invalid_variant: String,
+) -> Result<String, RoverClientError> {
+    let service_data = match response_data.service {
+        Some(data) => Ok(data),
+        None => Err(RoverClientError::NoService {
+            graph: graph.clone(),
+        }),
+    }?;
+
+    let mut valid_variants = Vec::new();
+
+    for variant in service_data.variants {
+        valid_variants.push(variant.name)
+    }
+
+    if service_data.schema.is_none() {
+        return Err(RoverClientError::NoSchemaForVariant {
+            graph,
+            invalid_variant,
+            valid_variants,
+            frontend_url_root: response_data.frontend_url_root,
+        });
+    }
+
+    if let Some(composition_publish) = service_data.most_recent_composition_publish {
+        if let Some(supergraph_sdl) = composition_publish.supergraph_sdl {
+            Ok(supergraph_sdl)
+        } else {
+            let mut error_msg = "".to_string();
+            // TODO: make this less hacky and let folks know these are composition errors.
+            for error in composition_publish.errors {
+                error_msg.push_str(&error.message);
+                error_msg.push_str("\n")
+            }
+            Err(RoverClientError::AdhocError { msg: error_msg })
+        }
+    } else {
+        Err(RoverClientError::NoCompositionPublishes { graph })
+    }
+}
```

Comment:
attn reviewers: this here is the crux of what i'm talking about - this seemed like a reasonable approach based on what the api gave me, but i'm really not sure if this is correct or exactly the behavior we want.